src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,468 @@
+        if object_version:
+            parent_param = f', {_ObjectField.PARENT_NAME}:"{object_name}"'
+            query_object_name = object_version
+
+        query_template = '{{{nodeKey}({domainKey}: {domain}, {nameKey}:"{name}"{parent_param}) {{{edges}}}}}'
+        query = query_template.format(
+            nodeKey=_DGQLFields.NODE,
+            domainKey=_ObjectField.DOMAIN,
+            nameKey=_ObjectField.NAME,
+            domain=object_domain.upper(),
+            name=query_object_name,
+            parent_param=parent_param,
+            edges="".join(parts),
+        )
+
+        return f"select SYSTEM$DGQL('{query}')"
+
+    def _get_lineage(
+        self,
+        object_name: str,
+        object_domain: str,
+        directions: list,
+        object_version: Optional[str] = None,
+        current_depth=1,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Constructs and executes a query to trace the lineage of a given entity at a single depth level.
+        """
+        query_string = self._build_graphql_query(
+            object_name, object_domain, directions, object_version
+        )
+        response = self._session.sql(query_string)
+        json_response = json.loads(response.collect()[0][0])
+
+        rows = []
+        for direction in directions:
+            edges = (
+                json_response.get(_DGQLFields.DATA, {})
+                .get(_DGQLFields.NODE, {})
+                .get(direction.value, [])
+            )
+
+            for edge in edges:
+                if _DGQLFields.SOURCE in edge and _DGQLFields.TARGET in edge:
+                    rows.append(
+                        (
+                            edge[_DGQLFields.SOURCE],
+                            edge[_DGQLFields.TARGET],
+                            direction,
+                            current_depth,
+                        )
+                    )
+
+        return rows
+
+    def _recursive_trace(
+        self,
+        object_name: str,
+        object_domain: str,
+        direction: str,
+        current_depth: int,
+        total_depth: int,
+        object_version: Optional[str] = None,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Recursively traces lineage by making successive DGQL queries based on response nodes.
+        """
+        if current_depth > total_depth:
+            return []
+
+        lineage_edges = self._get_lineage(
+            object_name, object_domain, [direction], object_version, current_depth
+        )
+        if (
+            current_depth == total_depth
+            or not lineage_edges
+            or self._is_terminal_entity(lineage_edges[0][0])
+            or self._is_terminal_entity(lineage_edges[0][1])
+        ):
+            return lineage_edges
+
+        next_object = (
+            lineage_edges[0][1]
+            if direction == LineageDirection.DOWNSTREAM
+            else lineage_edges[0][0]
```

Comment:
why are we just getting one object for next_object, wouldn't there be a list of next_objects? for cases like A -> [B, C, D], we need to call recursively on B, C, D separately. Also if I am understanding correctly, we are doing DFS, we can hit the same object multiple times in the graph, are we doing optimization to avoid calling DGQL on the object multiple times? for example A -> [B, C -> [B, D], D], and C's downstream B/D might already be visited

Reply:
Output will have list of edges (source, target). I am processing target in each edge. 
 I am tracking the visited nodes to avoid loops.

Reply:
I swicthed the logic to BFS. PTAL